src/cli/ops/kernels.c

```diff
@@ -0,0 +1,166 @@
+                } else {
+                        boot_manager_set_image_mode(manager, forced_image);
+                }
+
+                /* CBM will check this again, we just needed to check for
+                 * image mode.. */
+                if (!boot_manager_set_prefix(manager, root)) {
+                        return false;
+                }
+        } else {
+                boot_manager_set_image_mode(manager, forced_image);
+                /* Default to "/", bail if it doesn't work. */
+                if (!boot_manager_set_prefix(manager, "/")) {
+                        return false;
+                }
+        }
+
+        /* Let CBM take care of the rest */
+        kernels = boot_manager_list_kernels(manager);
+        if (!kernels) {
+                return false;
+        }
+        for (char **k = kernels; *k; k++) {
+                printf("%s\n", *k);
+                free(*k);
+        }
+        free(kernels);
+        return true;
+}
+
+bool cbm_command_set_kernel(int argc, char **argv)
+{
+        autofree(char) *root = NULL;
+        autofree(BootManager) *manager = NULL;
+        bool forced_image = false;
+        char type[32] = { 0 };
+        char version[16] = { 0 };
+        int release = 0;
+        Kernel kern = { 0 };
+
+        if (!cli_default_args_init(&argc, &argv, &root, &forced_image)) {
+                return false;
+        }
+
+        manager = boot_manager_new();
+        if (!manager) {
+                DECLARE_OOM();
+                return false;
+        }
+
+        if (root) {
+                autofree(char) *realp = NULL;
+
+                realp = realpath(root, NULL);
+                if (!realp) {
+                        LOG_FATAL("Path specified does not exist: %s", root);
+                        return false;
+                }
+                /* Anything not / is image mode */
+                if (!streq(realp, "/")) {
+                        boot_manager_set_image_mode(manager, true);
+                } else {
+                        boot_manager_set_image_mode(manager, forced_image);
+                }
+
+                /* CBM will check this again, we just needed to check for
+                 * image mode.. */
+                if (!boot_manager_set_prefix(manager, root)) {
+                        return false;
+                }
+        } else {
+                boot_manager_set_image_mode(manager, forced_image);
+                /* Default to "/", bail if it doesn't work. */
+                if (!boot_manager_set_prefix(manager, "/")) {
+                        return false;
+                }
+        }
+
+        if (argc != 1) {
+                fprintf(stderr,
+                        "set-kernel takes a kernel ID of the form %s.TYPE.VERSION-RELEASE\n",
+                        KERNEL_NAMESPACE);
+                return false;
+        }
+
+        if (sscanf(argv[optind], KERNEL_NAMESPACE ".%31[^.].%15[^-]-%d", type, version, &release) != 3) {
```

Comment:
So why do you use `optind` here? It belongs to getopts which is not used explicitly here and is not extern'ed. Let's just use 0.